federationapi/routing/join.go

```diff
@@ -271,17 +271,27 @@ func SendJoin(
 
 	// Check if the user is already in the room. If they're already in then
 	// there isn't much point in sending another join event into the room.
+	// Also check to see if they are banned: if they are then we reject them.
 	alreadyJoined := false
+	isBanned := false
 	for _, se := range stateAndAuthChainResponse.StateEvents {
 		if !se.StateKeyEquals(*event.StateKey()) {
 			continue
 		}
 		if membership, merr := se.Membership(); merr == nil {
 			alreadyJoined = (membership == gomatrixserverlib.Join)
+			isBanned = (membership == gomatrixserverlib.Ban)
 			break
 		}
 	}
 
+	if isBanned {
+		return util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("user is banned"),
```

Comment:
If this is a message that will get passed through to clients then we might want to make it friendlier.

Reply:
Friendlier how? I'm saying they are banned. Also, this requires explicit fudginess for it to work (you need to make_join with an unbanned user, then swap the state_key over to a banned user), so I don't expect this to happen in normal operation.